File: 3.x.x/crates-utopia/dns/src/resolver/proxy.rs
use crate::client::Client;
use crate::error::Result;
use crate::query::Query;
use crate::resolver::Resolver;
use crate::Message;

/// Forwards queries to another DNS server. PHP `Utopia\DNS\Resolver\Proxy`.
#[derive(Debug)]
pub struct Proxy {
    client: Client,
}

impl Proxy {
    pub fn new(server: impl Into<String>, port: u16) -> Result<Self> {
        Ok(Self {
            client: Client::new(server, port, 5, false)?,
        })
    }
}

impl Resolver for Proxy {
    fn resolve(&self, query: &Query) -> Result<Message> {
        self.client.query(&query.message)
    }
}

File: packages/dns/src/DNS/Resolver/Proxy.php
<?php

declare(strict_types=1);

namespace Utopia\DNS\Resolver;

use Utopia\DNS\Client;
use Utopia\DNS\Message;
use Utopia\DNS\Query;
use Utopia\DNS\Resolver;

class Proxy implements Resolver
{
    protected Client $client;

    /**
     * Create a new Proxy resolver
     *
     * @param string $server DNS server IP address
     * @param int $port DNS server port (default: 53)
     */
    public function __construct(protected string $server, protected int $port = 53)
    {
        $this->client = new Client($this->server, $this->port);
    }

    /**
     * Resolve DNS Record by proxying to another DNS server
     */
    public function resolve(Query $query): Message
    {
        return $this->client->query($query->message);
    }
}
